File: src/main/java/fr/lunatech/timekeeper/timeutils/TimeSheetUtils.java
/*
 * Copyright 2020 Lunatech S.A.S
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.lunatech.timekeeper.timeutils;

import fr.lunatech.timekeeper.models.time.TimeSheet;

public class TimeSheetUtils {

    private static final int NUMBER_OF_HOURS_IN_A_DAY = 8;
    private static final int NUMBER_OF_HOURS_IN_HALF_A_DAY = 4;

    private TimeSheetUtils() {
    }

    public static Long computeLeftOver(TimeSheet sheet) {
        if (null != sheet.maxDuration && sheet.maxDuration > 0) {
            // Early exit if no entry : left = maxDuration
            if (sheet.entries == null || sheet.entries.isEmpty()) {
                return sheet.maxDuration.longValue();
            }

            long consumedHours = sheet.entries.stream()
                    .mapToLong(timeEntry -> timeEntry.numberOfHours)
                    .sum();

            // Compute what's left based on durationUnit type (see constant)
            if (sheet.durationUnit != null) {
                switch (sheet.durationUnit) {
                    case DAY:
                        return ((sheet.maxDuration * NUMBER_OF_HOURS_IN_A_DAY) - consumedHours) / NUMBER_OF_HOURS_IN_A_DAY;
                    case HALFDAY:
                        return ((sheet.maxDuration * NUMBER_OF_HOURS_IN_HALF_A_DAY) - consumedHours) / NUMBER_OF_HOURS_IN_A_DAY;
                    case HOURLY:
                        return (sheet.maxDuration - consumedHours) / NUMBER_OF_HOURS_IN_A_DAY;
                    default:
                        throw new IllegalStateException("Invalid duration unit for sheet=" + sheet);
                }
            } else {
                // if no durationUnit let's consider it's a day based duration
                return ((sheet.maxDuration * NUMBER_OF_HOURS_IN_A_DAY) - consumedHours) / NUMBER_OF_HOURS_IN_A_DAY;
            }
        }
        return null;
    }
}

File: frontend/src/utils/timesheetUtils.js
/*
 * Copyright 2020 Lunatech S.A.S
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {getHalfDayDuration} from './configUtils';
import moment from 'moment';

// Filter the list based on start date end date and Given date
export const filteredProjects = (timeSheets, givenDate) => {
  return timeSheets.filter(timesheet =>
    (null == timesheet.startDate && null == timesheet.expirationDate) ||
      (moment(timesheet.startDate).isSameOrBefore(givenDate) && null == timesheet.expirationDate) ||
      (null == timesheet.startDate && moment(timesheet.expirationDate).isSameOrAfter(givenDate)) ||
      (moment(timesheet.startDate).isSameOrBefore(givenDate) && moment(timesheet.expirationDate).isSameOrAfter(givenDate))
  );
};

// Returns true if the timeUnits are disabled or if the date is out of range
export const isTimeSheetDisabled = (timeSheet, date, numberOfHoursForDay, entryDuration) => {
  if(timeSheet.leftOver === 0) return true; // days left equals 0
  return isTimeUnitDisabled(timeSheet, numberOfHoursForDay, entryDuration) || isDateOutOfTimeSheetRange(timeSheet, date);
};

// Returns true if all units are disabled, entryDuration is optional and is present only on timeEntry edition
const isTimeUnitDisabled = (timeSheet, numberOfHoursForDay, entryDuration) => {
  const timeUnit = timeSheet && timeSheet.timeUnit;
  const hourDisabled = timeUnit && timeUnit !== 'HOURLY';
  const halfDayDisabled = (timeUnit && timeUnit !== 'HOURLY' && timeUnit !== 'HALFDAY') ||
      (numberOfHoursForDay && entryDuration ? (numberOfHoursForDay - entryDuration) > getHalfDayDuration() : numberOfHoursForDay > getHalfDayDuration());
  const dayDisabled = numberOfHoursForDay && entryDuration ? (numberOfHoursForDay - entryDuration) > 0 : numberOfHoursForDay > 0;

  return hourDisabled && halfDayDisabled && dayDisabled;
};

// Returns true if the date is not in timesheets date range
const isDateOutOfTimeSheetRange = (timeSheet, date) => {
  if(timeSheet.expirationDate){
    return isDateOutOfTimeSheetRangeWithEndDate(timeSheet, date);
  }
  return isDateOutOfTimeSheetRangeWithOutEndDate(timeSheet, date);
};

// Returns true if the date is before the startDate with no endDate
const isDateOutOfTimeSheetRangeWithOutEndDate = (timeSheet, date) => {
  const startDate = moment(timeSheet.startDate);
  const dateFormatted = moment(date.format('YYYY-MM-DD'));
  return !(dateFormatted.isSameOrAfter(startDate));
};

// Returns true if the date is before the startDate or after the endDate
const isDateOutOfTimeSheetRangeWithEndDate = (timeSheet, date) => {
  const startDate = moment(timeSheet.startDate);
  const endDate = moment(timeSheet.expirationDate);
  const dateFormatted = moment(date.format('YYYY-MM-DD'));
  return !dateFormatted.isBetween(startDate, endDate,undefined, []);
};